### PR TITLE
Cluster Down/Remove

### PR DESCRIPTION
**Pull request by galal-hussein**
* Remove services' docker containers
* Clean up static mounted directories [/etc/kubernetes/ssl, /etc/cni, /opt/cni, /var/lib/etcd]

### CODE REVIEW

File: cluster/down.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package cluster
+
+import (
+	"github.com/rancher/rke/hosts"
+	"github.com/rancher/rke/services"
+)
+
+func (c *Cluster) ClusterDown() error {
```
Comment (moelsayed) [inline cluster/down.go]: We probably should tear down stuff in the revers order they were created.

File: cmd/cluster.go [hunk 2]
```diff
@@ -14,45 +14,43 @@ import (
...
 		Subcommands: []cli.Command{
 			cli.Command{
 				Name:   "up",
 				Usage:  "Bring the cluster up",
 				Action: clusterUpFromCli,
-				Flags:  clusterUpFlags,
+				Flags:  clusterUpDownUpgradeFlags,
+			},
+			cli.Command{
+				Name:   "down",
```
Comment (moelsayed) [inline cmd/cluster.go]: We should ask the user for confirmation before removal..We also should provide a --force option to skip the confirmation and allow for automation and shoot-self-in-the-foot needs..

File: docker/docker.go [hunk 3]
```diff
@@ -69,6 +69,32 @@ func DoRollingUpdateContainer(dClient *client.Client, imageCfg *container.Config
 	return err
 }
 
+func DoRemoveContainer(dClient *client.Client, containerName, hostname string) error {
+	logrus.Infof("[down/%s] Checking if container is running on host [%s]", containerName, hostname)
+	// not using the wrapper to check if the error is a NotFound error
+	_, err := dClient.ContainerInspect(context.Background(), containerName)
```
Comment (moelsayed) [inline docker/docker.go]: Why not use IsContainerRunning() ? or at least  InspectContainer() ?

File: services/kubeapi.go [hunk 4]
```diff
@@ -43,6 +43,10 @@ func upgradeKubeAPI(host hosts.Host, etcdHosts []hosts.Host, kubeAPIService v1.K
 	return err
 }
 
+func removeKubeAPI(host hosts.Host) error {
+	return docker.DoRemoveContainer(host.DClient, KubeAPIContainerName, host.AdvertisedHostname)
```
Comment (moelsayed) [inline services/kubeapi.go]: Add logging to this and the similar ones ? Like you have with etcd ?

Comment (galal-hussein) [reply services/kubeapi.go]: Actually i do in https://github.com/rancher/rke/pull/23/files#diff-a1440968342b4da00653d6c74b21f48eR56 and in workerplane file

File: docker/docker.go (hunk 3 above)
Comment (galal-hussein) [reply docker/docker.go]: This is because i need to inspect the error to make sure its not a notfound error

File: cluster/down.go (hunk 1 above)
Comment (galal-hussein) [reply cluster/down.go]: Done

File: cmd/cluster.go (hunk 2 above)
Comment (galal-hussein) [reply cmd/cluster.go]: I added force flag

**Comment by moelsayed** [comment]
LGTM.

**Pull request merged**